### PR TITLE
Use the right sized byte array for large indexes

### PR DESCRIPTION
**Pull request by Gouthamve**
Fixes: https://github.com/prometheus/prometheus/issues/3957

Not happy with this, suggestions welcome. @Bplotka

---
This change is [Reviewable](https://reviewable.io/reviews/prometheus/tsdb/301)

### CODE REVIEW

File: fileutil/mmap_windows.go [hunk 1]
```diff
@@ -35,7 +35,17 @@ func mmap(f *os.File, sz int) ([]byte, error) {
...
-	return (*[1 << 30]byte)(unsafe.Pointer(addr))[:sz], nil
+	if sz < 1<<30 {
+		return (*[1 << 30]byte)(unsafe.Pointer(addr))[:sz], nil
+	}
+
+	if sz < 1<<32 {
+		return (*[1 << 32]byte)(unsafe.Pointer(addr))[:sz], nil
+	}
+
+	if sz <= 1<<34 {
```
Comment (brian-brazil) [inline fileutil/mmap_windows.go]: Why not always use this?

Comment (Gouthamve) [reply fileutil/mmap_windows.go]: You are allocating 16G everytime? Whether or not we are using it.

Comment (brian-brazil) [reply fileutil/mmap_windows.go]: Are you sure that's an allocation?

File: fileutil/mmap_windows.go [hunk 2]
```diff
@@ -35,7 +35,7 @@ func mmap(f *os.File, sz int) ([]byte, error) {
 		return nil, os.NewSyscallError("CloseHandle", err)
 	}
 
-	return (*[1 << 30]byte)(unsafe.Pointer(addr))[:sz], nil
```
Comment (Bplotka) [inline fileutil/mmap_windows.go]: `sz` is just `size`? Can we name it like that? (maybe just for master)

Comment (Bplotka) [reply fileutil/mmap_windows.go]: Also I have stupid question. I have never used `mmap` before, but... why don't we cast to just `[sz]byte` here?

Comment (krasi-georgiev) [reply fileutil/mmap_windows.go]: I was thinking the same thing and tried it but casting to an array requires a constant for the array size as this is evaluated at compile time

-	return (*[sz]byte)(unsafe.Pointer(addr))[:sz], nil	=  non-constant array bound

**Comment by fabxc** [comment]
👍 

tsdb needs another revendor then

**Pull request merged**